pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,248 @@
+	descriptorsByMessageName map[string]model.ProtoSchema
+	logger                   logger
+}
+
+func NewController(logger logger) CoreDataModel {
+	descriptorsByMessageName := make(map[string]model.ProtoSchema, len(model.IstioConfigTypes))
+	for _, config := range model.IstioConfigTypes {
+		descriptorsByMessageName[config.MessageName] = config
+	}
+
+	configStore := make(map[string]map[string]*sync.Map)
+	for _, typ := range model.IstioConfigTypes.Types() {
+		configStore[typ] = make(map[string]*sync.Map)
+	}
+	return &Controller{
+		configStore:              configStore,
+		eventCh:                  make(chan func(model.Config, model.Event)),
+		updateCh:                 make(chan changeEvent),
+		descriptorsByMessageName: descriptorsByMessageName,
+		logger: logger,
+	}
+}
+
+func (c *Controller) Run(stop <-chan struct{}) {
+	for {
+		select {
+		case <-stop:
+			if _, ok := <-c.eventCh; ok {
+				close(c.eventCh)
+			}
+			return
+		case event, ok := <-c.eventCh:
+			if ok {
+				// TODO: do we need to call the events with any other args?
+				event(model.Config{}, model.EventUpdate)
+			}
+		case change, ok := <-c.updateCh:
+			if ok {
+				conf, exists := c.Get(change.descriptorType, change.metaName, "")
+				if exists {
+					conf.Spec = change.resource
+					// handle the error by inserting it in a chan
+					c.Update(*conf)
+				}
+			}
+		}
+	}
+}
+
+func (c *Controller) RegisterEventHandler(typ string, handler func(model.Config, model.Event)) {
+	c.eventCh <- handler
+}
+
+func (c *Controller) ConfigDescriptor() model.ConfigDescriptor {
+	return model.IstioConfigTypes
+}
+
+func (c *Controller) Get(typ, name, namespace string) (*model.Config, bool) {
+	_, ok := c.configStore[typ]
+	if !ok {
+		c.logger.Infof("Get: config not found for the type %s", typ)
+		return nil, false
+	}
+
+	ns, exists := c.configStore[typ][namespace]
+	if !exists {
+		c.logger.Infof("Get: config not found for the type %s", typ)
+		return nil, false
+	}
+
+	out, exists := ns.Load(name)
+	if !exists {
+		return nil, false
+	}
+	config := out.(model.Config)
+
+	return &config, true
+}
+
+func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+	_, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("List: unknown type %s", typ))
+	}
+	data, exists := c.configStore[typ]
+	if !exists {
+		c.logger.Infof("List: config not found for the type %s", typ)
+		return nil, nil
+	}
+	out := make([]model.Config, 0, len(c.configStore[typ]))
+	if namespace == "" {
+		for _, ns := range data {
+			ns.Range(func(key, value interface{}) bool {
+				out = append(out, value.(model.Config))
+				return true
+			})
+		}
+	} else {
+		ns, exists := data[namespace]
+		if !exists {
+			return nil, nil
+		}
+		ns.Range(func(key, value interface{}) bool {
+			out = append(out, value.(model.Config))
+			return true
+		})
+	}
+	return out, nil
+}
+
+func (c *Controller) Update(config model.Config) (newRevision string, err error) {
```

Comment:
Update and Create - I think you can remove, no need to keep the code and test it. 
We're not planning to use them ( will be removed from model probably pretty soon )

Reply:
Sure that makes sense, however the create is used to populate some test data during tests, otherwise without create it wouldn't be easy to test since this controller does not allow you to inject a configStore nor it exposes it.

Reply:
I would use a fake MCP server ( it's just a grpc service ). Long term the MCP client should not hold state - that is a problem for the current k8s adapter ( caches more than we need and not what we need ).

Reply:
Add a comment that it's only used for testing. I would use a dedicated SetStoreForTest method - or keep the store in the fake MCP server. As I mentioned, the controller should not have a store inside long term.

Reply:
https://github.com/istio/istio/blob/32967f9222f350faac3aa029ec6a3df1bde489bd/pkg/mcp/testing/server.go#L31